### PR TITLE
Fix assembly item comparison for verbatim.

### LINKED ISSUE #14640: In yul, there's a bug when mixing `verbatim`, `switch`, and unoptimized code
## Description

We need unoptimized Yul code to measure gas costs (see https://github.com/ethereum/solidity/issues/14621). This is often facilitated using expressions such as:

```
delta := verbatim_1i_1o(hex"<gas><swap1><operation, which uses the parameter><gas><swap1><sub>", <input value>)
```

However, such expressions tend to confuse the compiler when put inside a `switch` statement

## Environment

- Compiler version:
  ```
  qbzzt1@tests:~/yul$ ./solc --version
  solc, the solidity compiler commandline interface
  Version: 0.8.21+commit.d9974bed.Linux.g++
  ```

- Target EVM version (as per compiler settings): The default, but I also verified with `--evm-version shanghai`.

- Framework/IDE (e.g. Truffle or Remix): None

- EVM execution environment / backend / blockchain client: Geth, but I can prove that's not the issue.

- Operating system: 

  Running on GCP
  ```
  qbzzt1@tests:~/yul$ uname -a
  Linux tests 5.10.0-26-cloud-amd64 #1 SMP Debian 5.10.197-1 (2023-09-29) x86_64 GNU/Linux
  qbzzt1@tests:~/yul$ 
  ```

## Steps to Reproduce

Compile this Yul code:

```yul
{
  let dummy := 0xAABBCCDDEEFF
  let input := sload(0)
  let output

  // I also verified this when switching on a variable's value
  switch input
  // This is the case that needs to be executed
  case 0x00 {
      output := verbatim_1i_1o(hex"506000", dummy)

      // Case gets executed with this (logically equivalent) code:
      // output := verbatim_0i_1o(hex"6000")
  }
  case 0x01 {
      output := 1 
  }
  // This is the case that actually gets executed.
  // It is the last but one case.
  case 0x02 {
      output := verbatim_1i_1o(hex"506002", dummy)

      // The first case gets executed correctly if I put here
      // this (logically equivalent) code
      // output := verbatim_0i_1o(hex"6002")
  }          
  case 0x03 {
      output := 3
  }          

  sstore(0, output)
}
```

Using the command line:

```sh
./solc --strict-assembly --optimize --yul-optimizations=":" x.yul 
```

The binary you get is: 

```
5f805465aabbccddeeff8115602e578160011460385781600214602e57506003146027575f55005b5060035f55005b9150505060025f55005b50505060015f5500
```

When I run it on geth, starting with empty storage, I get this output:

```
    "0xcccccccccccccccccccccccccccccccccccccccc" : {
        "code" : "0x5f805465aabbccddeeff8115602e578160011460385781600214602e57506003146027575f55005b5060035f55005b9150505060025f55005b50505060015f5500",
        "nonce" : "0x01",
        "balance" : "0x0186a0",
        "storage" : {
            "0x00" : "0x02"
        }
    }
```

Even though `sload(0)` returns zero (we start from an empty storage), so `input` is zero and therefore we'd expect the first case of the `switch` to happen, not the third.

Here is a raw trace:

```
{"pc":0,"op":95,"gas":"0x26207f8","gasCost":"0x2","memSize":0,"stack":[],"depth":1,"refund":0,"opName":"PUSH0"}
{"pc":1,"op":128,"gas":"0x26207f6","gasCost":"0x3","memSize":0,"stack":["0x0"],"depth":1,"refund":0,"opName":"DUP1"}
{"pc":2,"op":84,"gas":"0x26207f3","gasCost":"0x834","memSize":0,"stack":["0x0","0x0"],"depth":1,"refund":0,"opName":"SLOAD"}
{"pc":3,"op":101,"gas":"0x261ffbf","gasCost":"0x3","memSize":0,"stack":["0x0","0x0"],"depth":1,"refund":0,"opName":"PUSH6"}
{"pc":10,"op":129,"gas":"0x261ffbc","gasCost":"0x3","memSize":0,"stack":["0x0","0x0","0xaabbccddeeff"],"depth":1,"refund":0,"opName":"DUP2"}
{"pc":11,"op":21,"gas":"0x261ffb9","gasCost":"0x3","memSize":0,"stack":["0x0","0x0","0xaabbccddeeff","0x0"],"depth":1,"refund":0,"opName":"ISZERO"}
{"pc":12,"op":96,"gas":"0x261ffb6","gasCost":"0x3","memSize":0,"stack":["0x0","0x0","0xaabbccddeeff","0x1"],"depth":1,"refund":0,"opName":"PUSH1"}
{"pc":14,"op":87,"gas":"0x261ffb3","gasCost":"0xa","memSize":0,"stack":["0x0","0x0","0xaabbccddeeff","0x1","0x2e"],"depth":1,"refund":0,"opName":"JUMPI"}
{"pc":46,"op":91,"gas":"0x261ffa9","gasCost":"0x1","memSize":0,"stack":["0x0","0x0","0xaabbccddeeff"],"depth":1,"refund":0,"opName":"JUMPDEST"}
{"pc":47,"op":145,"gas":"0x261ffa8","gasCost":"0x3","memSize":0,"stack":["0x0","0x0","0xaabbccddeeff"],"depth":1,"refund":0,"opName":"SWAP2"}
{"pc":48,"op":80,"gas":"0x261ffa5","gasCost":"0x2","memSize":0,"stack":["0xaabbccddeeff","0x0","0x0"],"depth":1,"refund":0,"opName":"POP"}
{"pc":49,"op":80,"gas":"0x261ffa3","gasCost":"0x2","memSize":0,"stack":["0xaabbccddeeff","0x0"],"depth":1,"refund":0,"opName":"POP"}
{"pc":50,"op":80,"gas":"0x261ffa1","gasCost":"0x2","memSize":0,"stack":["0xaabbccddeeff"],"depth":1,"refund":0,"opName":"POP"}
{"pc":51,"op":96,"gas":"0x261ff9f","gasCost":"0x3","memSize":0,"stack":[],"depth":1,"refund":0,"opName":"PUSH1"}
{"pc":53,"op":95,"gas":"0x261ff9c","gasCost":"0x2","memSize":0,"stack":["0x2"],"depth":1,"refund":0,"opName":"PUSH0"}
{"pc":54,"op":85,"gas":"0x261ff9a","gasCost":"0x4e20","memSize":0,"stack":["0x2","0x0"],"depth":1,"refund":0,"opName":"SSTORE"}
{"pc":55,"op":0,"gas":"0x261b17a","gasCost":"0x0","memSize":0,"stack":[],"depth":1,"refund":0,"opName":"STOP"}
{"output":"","gasUsed":"0x567e"}
```

Tracing the code manually shows us what happens:

| PC | Opcode | Stack (after operation) |
| -: | - | :- |
| 0 | PUSH0 | 0
| 1 | DUP1 | 0,0
| 2 | SLOAD | 0,0
| 3 | PUSH6 0xaabbccddeeff | 0xaa..ff, 0, 0
| 10 | DUP2 | 0, 0xaa..ff, 0, 0
| 11 | ISZERO | 1, 0xaa..ff, 0, 0
| 12 | PUSH1 0x2E | 0x2E, 1, 0xaa..ff, 0, 0
| 14 | JUMPI | 0xaa..ff, 0, 0
| (0x2e=) 46 | JUMPDEST | 0xaa..ff, 0, 0
| 47 | SWAP2 | 0, 0, 0xaa..ff
| 48 | POP | 0, 0xaa..ff
| 49 | POP |  0xaa..ff
| 50 | POP | 
| 51 | PUSH1 0x02 | 2
| 53 | PUSH0 | 0, 2
| 54 | SSTORE | 
| 55 | STOP |

Note that this is an edge case. If I replace either of the `verbatim` expressions with `verbatim_0i_1o(hex"<same operation, but without the initial pop>")`, the compiler works correctly.

### PR DESCRIPTION
**Pull request by ekpyron**
This is the cause of and thereby fixes https://github.com/ethereum/solidity/issues/14640.

It's present ever since the introduction of ``verbatim`` in https://github.com/ethereum/solidity/pull/11123.

The hard question here will be to evaluate what effects this can possibly have and thereby what severity to assign to it.

### CODE REVIEW

File: docs/bugs.json
Comment (ekpyron) [inline docs/bugs.json]:
```suggestion
        "name": "VerbatimInvalidDeduplication",
        "summary": "``verbatim`` calls with different data can wrongfully be de-duplicated if used in very similar context.",
        "description": "The block deduplicator is an optimizer step which identifies equivalent assembly blocks and merges them into a single one. However when blocks contained ``verbatim``, their comparison was implemented incorrectly, leading to the collapse of assembly blocks which are identical except on the contents of the ``verbatim`` items. Since ``verbatim`` is only available in Yul, compilation of Solidity sources is not affected.",
```

File: docs/bugs.json
Comment (ekpyron) [inline docs/bugs.json]: Not entirely sure if ``very low`` wouldn't even be sufficient. Another case where it'd be much easier if we split this into likelihood and impact :-). (Likelihood is near zero, while impact is actually high)

File: docs/bugs.json
Comment (ekpyron) [inline docs/bugs.json]:
```suggestion
```
The block deduplicator is part of the libevmasm optimizer, so the yul optimizer doesn't need to be enabled for this.

File: docs/bugs.json
Comment (ekpyron) [reply docs/bugs.json]: Call decision: we keep it at low.

File: Changelog.md
Comment (matheusaaguiar) [inline Changelog.md]:
```suggestion
 * Yul Optimizer: Fix block deduplicator bug which allowed that blocks which are identical but have different ``verbatim`` contents be treated as equivalent and thus collapsed into a single one.
```

File: Changelog.md
Comment (cameel) [inline Changelog.md]: This one should go into `Important Bugfixes:` and the top.

Also, this is not a part of the Yul optimizer. It's in `libevmasm` and is controlled by `settings.optimizer.deduplicate`. I see that there's one previous entry about deduplicator and we filed that under `Optimizer`.

But it would probably still a good idea to indicate in some way that this only affects assembling of Yul code.

File: docs/bugs.json
Comment (cameel) [inline docs/bugs.json]: Sounds ok (other than it should be "wrongly", not "wrongfully"), but maybe that would be simpler to understand to users?

```suggestion
        "summary": "All ``verbatim`` blocks are considered identical by deduplicator and can wrongly be unified when surrounded by identical opcodes.",
```

File: docs/bugs.json
Comment (cameel) [inline docs/bugs.json]: Just a few small tweaks:

```suggestion
        "description": "The block deduplicator is a step of the opcode-based optimizer which identifies equivalent assembly blocks and merges them into a single one. However, when blocks contained ``verbatim``, their comparison was performed incorrectly, leading to the collapse of assembly blocks which are identical except for the contents of the ``verbatim`` items. Since ``verbatim`` is only available in Yul, compilation of Solidity sources is not affected.",
```

File: Changelog.md
Comment (matheusaaguiar) [reply Changelog.md]: Done. I added a line mentioning that.

File: docs/bugs.json
Comment (matheusaaguiar) [reply docs/bugs.json]: I guess "incorrectly" is better?

File: Changelog.md
Comment (cameel) [inline Changelog.md]:
```suggestion
 * Optimizer: Fix block deduplicator bug which allowed that blocks which are identical but have different ``verbatim`` contents be treated as equivalent and thus collapsed into a single one. Only assembling of Yul code was affected.
```

**Review by cameel — COMMENTED** [review]
Seems fine now.

File: Changelog.md
Comment (ekpyron) [inline Changelog.md]:
```suggestion
 * Optimizer: Fix block deduplicator bug which led to blocks which are identical apart from the contents of ``verbatim`` instructions to be treated as equivalent and thus collapsed into a single one.
```

A bit nitpicky, but still :-).

**Comment by ekpyron** [comment]
In principle, a test would be nice :-). Rather surprised that @cameel didn't cry out for one :-).
We can merge it without if we don't have the time before the release, but would still be nice to have.

**Comment by cameel** [comment]
Yeah, definitely should have a test. How did I miss that :P

**Comment by ekpyron** [comment]
I'm not entirely sure what kind of test would work best here. Maybe an ``test/libyul/objectCompiler`` test with an enabled optimizer preset and a switch like in https://github.com/ethereum/solidity/issues/14640 should do it?

The test needs to start from something with ``verbatim``, so Yul... and run compilation down to evmasm assembly through the libevmasm optimizer... I think apart from the `objectCompiler` tests, it'd have to be a command line test - but ``test/libyul/objectCompiler`` with `optimizationPreset : standard` should work!

**Comment by cameel** [comment]
We could do what I did in #14574 and [create a script-based CLI test](https://github.com/ethereum/solidity/pull/14574/files#diff-15e1a6fbee03546d8c00bd79fe5e503bc169bca629a2c7b6bc28fa837e02899c). The test could have an example with two different patterns in `verbatim` (long enough to be unlikely to occur naturally) and grep the bytecode to check that both are present.

I guess an `objectCompiler` test would work too, though with a little higher risk that when updating it in the future (due to unrelated bytecode changes) we won't even notice when the output no longer is what it should be. But this kind of test has the advantage that it shows you the assembly, which is why in #14574 I included one anyway.

File: test/libyul/objectCompiler/verbatim_bug.yul
Comment (ekpyron) [inline test/libyul/objectCompiler/verbatim_bug.yul:115]: Note that in the first commit that adds this test case, there is no occurrence of ``verbatimbytecode_506000`` at all, while it reappears in the second commit with the fix.

**Comment by ekpyron** [comment]
I can't approve this, since I'm the original author - but let this comment count as an approval and feel free to approve and merge once CI passes.

File: test/libyul/objectCompiler/verbatim_bug.yul
Comment (ekpyron) [inline test/libyul/objectCompiler/verbatim_bug.yul]: The bug can actually be seen here directly in the second commit - here the invalid replacement of tags due to the bug is reversed. Of course the block pointed to by the tag also reappears but its removal before the bugfix was not actually part of the bug (after the buggy tag replacement the removal of that blog is perfectly valid optimizer behaviour, since due to the bug it had become an unused block).

File: test/cmdlineTests/~deduplicator-verbatim-bug/test.sh
Comment (r0qs) [inline test/cmdlineTests/~deduplicator-verbatim-bug/test.sh:19]: LGTM. Just a nitpick, for arithmetic variables we usually use the following form:
```suggestion
(( VERBATIM_OCCURRENCES == 2 )) || assertFail "Incorrect number of verbatim items in assembly."
```

**Review by r0qs — APPROVED** [review]

**Pull request merged**
